packages/siwe/src/provider.tsx

```diff
@@ -0,0 +1,120 @@
+import { useEthers } from '@usedapp/core'
+import React, { useEffect, ReactNode, useState } from 'react'
+import { createContext, useContext } from 'react'
+import { SiweMessage } from 'siwe'
+
+export interface SiweContextInterface {
+  signIn: () => void
+  signOut: () => void
+  isLoggedIn: boolean
+}
+
+const SiweContext = createContext<SiweContextInterface>({
+  signIn: () => undefined,
+  signOut: () => undefined,
+  isLoggedIn: false,
+})
+
+export function useSiwe() {
+  return useContext(SiweContext)
+}
+
+export interface SiweProviderProps {
+  children: ReactNode
+  backendUrl: string
+}
+
+export const SiweProvider = ({ children, backendUrl }: SiweProviderProps) => {
+  const { account, chainId, library } = useEthers()
+  const [isLoggedIn, setLoggedIn] = useState(false)
+  const [address, setAddress] = useState<string | undefined>(undefined)
+  const [sessionChainId, setSessionChainId] = useState<number | undefined>(undefined)
+
+  useEffect(() => {
+    if ((!!account && account !== address) || (!!chainId && chainId !== sessionChainId)) {
+      setLoggedIn(false)
+    }
+  }, [account, chainId])
+
+  useEffect(() => {
+    if (!account) {
+      return
+    }
+    const checkAuthStatus = async () => {
+      const authRequest = await fetch(`${backendUrl}/auth/me`, {
+        method: 'GET',
+        credentials: 'include',
+      })
+      const authResponse = await authRequest.json()
+
+      if (authRequest.ok && authResponse.address === account) {
+        setAddress(authResponse.address)
+        setLoggedIn(true)
+      }
+    }
+    checkAuthStatus()
+  }, [account, chainId])
+
+  const signIn = async () => {
+    if (!account || !chainId || !library) {
+      return
+    }
+    const nonceRequest = await fetch(`${backendUrl}/auth/siwe/nonce`, {
```

Comment:
extract all fetches to functions in a separate file that have their return types sepcified